lib/easybill_rest_client/request.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+require 'easybill_rest_client/request_logger'
+require 'easybill_rest_client/retry_on'
+
+module EasybillRestClient
+  class Request
+    BASE_URL = 'https://api.easybill.de/rest/v1'
+    USERNAME = 'rest-api@easybill.de'
+    SUPPORTED_METHODS = %i(get post put delete).freeze
+    OPEN_TIMEOUT = 5
+
+    def initialize(opts = {})
+      unless SUPPORTED_METHODS.include?(opts.fetch(:method).to_sym)
+        raise ArgumentError, "Unsupported HTTP method '#{method}'"
+      end
+
+      @api_key = opts.fetch(:api_key)
+      @method = opts.fetch(:method)
+      @endpoint = opts.fetch(:endpoint)
+      @params = opts.fetch(:params)
+      @logger = opts.fetch(:logger)
+      @tries = opts.fetch(:tries)
+      @retry_cool_off_time = opts.fetch(:retry_cool_off_time)
+    end
+
+    def run
+      request_logger.info("#{method.to_s.upcase} #{endpoint} #{request_details}")
+      retry_on(TooManyRequests) do
+        retry_on(Net::OpenTimeout) do
+          response = perform_request
+          raise TooManyRequests if response.is_a?(Net::HTTPTooManyRequests)
+          response
+        end
+      end
+    end
+
+    def request_details
+      body_allowed? ? request.body : uri.query
+    end
+
+    private
+
+    attr_reader :api_key, :method, :endpoint, :params, :logger, :request_logger, :retry_on, :tries,
+      :retry_cool_off_time
+
+    def perform_request
+      Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == 'https') do |http|
+        http.open_timeout = OPEN_TIMEOUT
```

Comment:
Those connections are by design not thread-safe. So please don't do this.

Here's some code to play around with:
```ruby
require 'net/http'

NUM = 1

c = Net::HTTP.new('www.google.com', 80)
t = NUM.times.map do
  Thread.new do
    5.times do
    puts c.get('/humans.txt')
    end
  end
end

t.each(&:join)
```


```
$ ruby x.rb
#<Net::HTTPOK:0x007fc41b0dab88>
#<Net::HTTPOK:0x007fc41b20af58>
#<Net::HTTPOK:0x007fc41b202178>
#<Net::HTTPOK:0x007fc41b1f95a0>
#<Net::HTTPOK:0x007fc41b1e9740>
```

But as soon as you change `NUM = 5` or similar, it will blow up in your face:

```
$ ruby x.rb
/Users/niko/.rbenv/versions/2.3.1/lib/ruby/2.3.0/net/http/response.rb:380:in `inflate': incorrect header check (Zlib::DataError)

$ ruby x.rb
#<Net::HTTPOK:0x007fcf7ab20c68>
/Users/niko/.rbenv/versions/2.3.1/lib/ruby/2.3.0/net/protocol.rb:158:in `wait_readable': closed stream (IOError)

$ ruby x.rb
#<Net::HTTPOK:0x007fe6f40d14d8>
x.rb: end of file reached (EOFError)
```

It's especially important to consider this already when writing code because usually it will work fine in production if you don't do a lot of requests. It will only sometimes fail and debugging this later is a huge pain.